// @sap-cloud-sdk/cli/lib/utils/jest-config.js
"use strict";
/*!
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.modifyJestConfig = exports.getJestConfig = void 0;
const fs = require("fs");
const utils_1 = require("../utils/");
function getJestConfig(isUnitTests) {
    return {
        reporters: [
            'default',
            [
                'jest-junit',
                {
                    suiteName: 'backend unit tests',
                    outputDirectory: `./s4hana_pipeline/reports/backend-${isUnitTests ? 'unit' : 'integration'}`
                }
            ]
        ],
        collectCoverage: true,
        coverageReporters: ['text', 'cobertura'],
        coverageDirectory: `../s4hana_pipeline/reports/coverage-reports/backend-${isUnitTests ? 'unit' : 'integration'}`
    };
}
exports.getJestConfig = getJestConfig;
function modifyJestConfig(jestConfigPath, data) {
    try {
        const jestConfig = JSON.parse(fs.readFileSync(jestConfigPath, {
            encoding: 'utf8'
        }));
        const adjustedJestConfig = Object.assign(Object.assign({}, jestConfig), data);
        fs.writeFileSync(jestConfigPath, JSON.stringify(adjustedJestConfig, null, 2));
    }
    catch (error) {
        utils_1.recordWarning(`Could not edit your Jest config at "${jestConfigPath}".`, 'Please verify if the location is correct and consider opening a bug ticket', 'at https://github.com/SAP/cloud-sdk-cli');
    }
}
exports.modifyJestConfig = modifyJestConfig;
