// @sap-cloud-sdk/cli/lib/commands/add-approuter.js
"use strict";
/*!
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const cli_ux_1 = require("cli-ux");
const Listr = require("listr");
const utils_1 = require("../utils/");
class AddApprouter extends command_1.Command {
    async run() {
        const { flags, args } = this.parse(AddApprouter);
        const projectDir = args.projectDir || '.';
        try {
            const options = await this.getOptions();
            const tasks = new Listr([
                {
                    title: 'Creating files',
                    task: async () => {
                        const copyDescriptors = utils_1.getCopyDescriptors(projectDir, utils_1.getTemplatePaths(['add-approuter']));
                        await utils_1.findConflicts(copyDescriptors, flags.force).catch(e => this.error(e, { exit: 11 }));
                        await utils_1.copyFiles(copyDescriptors, options);
                    }
                }
            ]);
            await tasks.run();
            this.printSuccessMessage();
        }
        catch (error) {
            this.error(error, { exit: 1 });
        }
    }
    async getOptions() {
        const projectName = utils_1.getProjectNameFromManifest(this);
        const options = {
            projectName: projectName || (await cli_ux_1.default.prompt('Enter project name as maintained in Cloud Foundry'))
        };
        return options;
    }
    printSuccessMessage() {
        this.log(`✅ Successfully added approuter to your project.

Generated files might need customization. Documentation available here:
- xs-security.json (for help check https://help.sap.com/viewer/4505d0bdaf4948449b7f7379d24d0f0d/2.0.02/en-US/e6fc90df44464a29952e1c2c36dd9861.html)
- xs-app.json (for help check https://help.sap.com/viewer/65de2977205c403bbc107264b8eccf4b/Cloud/en-US/c103fb414988447ead2023f768096dcc.html)
- mainfest.yml (for help check https://help.sap.com/viewer/65de2977205c403bbc107264b8eccf4b/Cloud/en-US/ba527058dc4d423a9e0a69ecc67f4593.html)`);
    }
}
exports.default = AddApprouter;
AddApprouter.description = 'Setup your Cloud Foundry app to authenticate through the app router';
AddApprouter.aliases = ['add-app-router'];
AddApprouter.examples = ['$ sap-cloud-sdk add-approuter'];
AddApprouter.flags = {
    force: command_1.flags.boolean({
        description: 'Do not fail if a file already exist and overwrite it.'
    }),
    help: command_1.flags.help({
        char: 'h',
        description: 'Show help for the add-approuter command.'
    })
};
AddApprouter.args = [
    {
        name: 'projectDir',
        description: 'Path to the project directory to which the approuter should be added.'
    }
];
