// @sap-cloud-sdk/core/dist/odata/v4/request-builder/response-data-accessor.js
"use strict";
/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSingleResult = exports.isCollectionResult = exports.getCollectionResult = void 0;
var util_1 = require("@sap-cloud-sdk/util");
var logger = util_1.createLogger({
    package: 'core',
    messageContext: 'response-data-accessor'
});
/**
 * Methods to extract the data from OData v4 responses.
 */
/**
 * Extract the collection data from the response.
 * If the data does not contain a collection an empty array is returned.
 * @param data - Response of the OData v4 service
 * @returns any[] - Collection extracted from the response
 */
function getCollectionResult(data) {
    validateCollectionResult(data);
    return isCollectionResult(data) ? data.value : [];
}
exports.getCollectionResult = getCollectionResult;
/**
 * Checks if the data contains a collection result.
 * @param data - Response of the OData v4 service
 * @returns boolean - true if the data is a collection result
 */
function isCollectionResult(data) {
    return Array.isArray(data.value);
}
exports.isCollectionResult = isCollectionResult;
function validateCollectionResult(data) {
    if (!isCollectionResult(data)) {
        logger.warn('The given reponse data does not have the standard OData v4 format for collections.');
    }
}
/**
 * Extract the single entry data from the response.
 * If the data does not contain a single object an empty object is returned.
 * @param data - Response of the OData v4 service
 * @returns MapType<any> - single object extracted from the response
 */
function getSingleResult(data) {
    validateSingleResult(data);
    return isSingleResult(data) ? data : {};
}
exports.getSingleResult = getSingleResult;
function isSingleResult(data) {
    return typeof data === 'object' && !Array.isArray(data);
}
function validateSingleResult(data) {
    if (!isSingleResult(data)) {
        logger.warn('The given reponse data does not have the standard OData v4 format for single results.');
    }
}
